File: final_report/cavity.py
"""
Reimplemented the jupyter notebook codes to Python3
Original codes: ../13_pde/10_cavity.ipynb
"""

from IPython.core.display import HTML
import numpy as np

from plot_cavity_flow import plot


def build_up_b(b, rho, dt, u, v, dx, dy):

    b[1:-1, 1:-1] = (rho * (1 / dt *
                            ((u[1:-1, 2:] - u[1:-1, 0:-2]) /
                             (2 * dx) + (v[2:, 1:-1] - v[0:-2, 1:-1]) / (2 * dy)) -
                            ((u[1:-1, 2:] - u[1:-1, 0:-2]) / (2 * dx))**2 -
                            2 * ((u[2:, 1:-1] - u[0:-2, 1:-1]) / (2 * dy) *
                                 (v[1:-1, 2:] - v[1:-1, 0:-2]) / (2 * dx)) -
                            ((v[2:, 1:-1] - v[0:-2, 1:-1]) / (2 * dy))**2))

    return b


def pressure_poisson(p, dx, dy, b):
    pn = np.empty_like(p)
    pn = p.copy()

    for q in range(nit):
        pn = p.copy()
        p[1:-1, 1:-1] = (((pn[1:-1, 2:] + pn[1:-1, 0:-2]) * dy**2 +
                          (pn[2:, 1:-1] + pn[0:-2, 1:-1]) * dx**2) /
                         (2 * (dx**2 + dy**2)) -
                         dx**2 * dy**2 / (2 * (dx**2 + dy**2)) *
                         b[1:-1, 1:-1])

        p[:, -1] = p[:, -2]  # dp/dx = 0 at x = 2
        p[0, :] = p[1, :]    # dp/dy = 0 at y = 0
        p[:, 0] = p[:, 1]    # dp/dx = 0 at x = 0
        p[-1, :] = 0         # p = 0 at y = 2

    return p


def cavity_flow(nt, u, v, dt, dx, dy, p, rho, nu):
    un = np.empty_like(u)
    vn = np.empty_like(v)
    b = np.zeros((ny, nx))

    for n in range(nt):
        un = u.copy()
        vn = v.copy()

        b = build_up_b(b, rho, dt, u, v, dx, dy)
        p = pressure_poisson(p, dx, dy, b)

        u[1:-1, 1:-1] = (un[1:-1, 1:-1] -
                         un[1:-1, 1:-1] * dt / dx *
                         (un[1:-1, 1:-1] - un[1:-1, 0:-2]) -
                         vn[1:-1, 1:-1] * dt / dy *
                         (un[1:-1, 1:-1] - un[0:-2, 1:-1]) -
                         dt / (2 * rho * dx) * (p[1:-1, 2:] - p[1:-1, 0:-2]) +
                         nu * (dt / dx**2 *
                               (un[1:-1, 2:] - 2 * un[1:-1, 1:-1] + un[1:-1, 0:-2]) +
                               dt / dy**2 *
                               (un[2:, 1:-1] - 2 * un[1:-1, 1:-1] + un[0:-2, 1:-1])))

        v[1:-1, 1:-1] = (vn[1:-1, 1:-1] -
                         un[1:-1, 1:-1] * dt / dx *
                         (vn[1:-1, 1:-1] - vn[1:-1, 0:-2]) -
                         vn[1:-1, 1:-1] * dt / dy *
                         (vn[1:-1, 1:-1] - vn[0:-2, 1:-1]) -
                         dt / (2 * rho * dy) * (p[2:, 1:-1] - p[0:-2, 1:-1]) +
                         nu * (dt / dx**2 *
                               (vn[1:-1, 2:] - 2 * vn[1:-1, 1:-1] + vn[1:-1, 0:-2]) +
                               dt / dy**2 *
                               (vn[2:, 1:-1] - 2 * vn[1:-1, 1:-1] + vn[0:-2, 1:-1])))

        u[0, :] = 0
        u[:, 0] = 0
        u[:, -1] = 0
        u[-1, :] = 1    # set velocity on cavity lid equal to 1
        v[0, :] = 0
        v[-1, :] = 0
        v[:, 0] = 0
        v[:, -1] = 0

    return u, v, p


if __name__ == '__main__':
    nx = 41
    ny = 41
    nt = 500
    nit = 50
    c = 1
    dx = 2 / (nx - 1)
    dy = 2 / (ny - 1)
    x = np.linspace(0, 2, nx)
    y = np.linspace(0, 2, ny)
    X, Y = np.meshgrid(x, y)

    rho = 1
    nu = .1
    dt = .001

    u = np.zeros((ny, nx))
    v = np.zeros((ny, nx))
    p = np.zeros((ny, nx))
    b = np.zeros((ny, nx))
    # Let's start with nt = 100 and see what the solver gives us
    nt = 100
    u, v, p = cavity_flow(nt, u, v, dt, dx, dy, p, rho, nu)
    result = {'u': u, 'v': v, 'p': p}

    plot(u, v, p, 'original_nt100.png')

    # You can see that two distinct pressure zones are forming and that the
    # spiral pattern expected from lid-driven cavity flow is beginning to form.
    # Experiment with different values of `nt` to see how long the system takes
    # to stabilize.
    u = np.zeros((ny, nx))
    v = np.zeros((ny, nx))
    p = np.zeros((ny, nx))
    b = np.zeros((ny, nx))
    nt = 700
    u, v, p = cavity_flow(nt, u, v, dt, dx, dy, p, rho, nu)

    plot(u, v, p, 'original_nt700.png')

File: final_report/cavity.cpp
#include <fstream>
#include <iostream>
#include <map>
#include <math.h>
#include <string>
#include <vector>

using namespace std;

int nx = 41;
int ny = 41;
int MAT_SIZE = nx * ny;
int nt = 700;
int nit = 50;
float c = 1.0;
float dx = 2.0 / (nx - 1);
float dy = 2.0 / (ny - 1);

int rho = 1.0;
float nu = 0.1;
float dt = 0.001;

void writeFile(float* u, float* v, float* p) {
  ofstream fs("cavity_cpp_results.txt");

  // u
  fs << "u ";
  for (int i = 0; i < MAT_SIZE; i++) fs << u[i] << " ";
  fs << "\n";
  fs << "v ";
  for (int i = 0; i < MAT_SIZE; i++) fs << v[i] << " ";
  fs << "\n";
  fs << "p ";
  for (int i = 0; i < MAT_SIZE; i++) fs << p[i] << " ";
  fs << "\n";

  fs.close();
}

void build_up_b(float *u, float *v, float *b) {
  for(int i = 1; i < ny - 1; i++){
    for(int j = 1; j < nx - 1; j++){
      b[i*nx+j] = (rho * ( 1.0 / dt *
            ((u[i*nx+j+1] - u[i*nx+j-1]) /
             (2.0 * dx) + (v[(i+1)*nx+j] - v[(i-1)*nx+j]) / (2.0 * dy)) -
            ((u[i*nx+j+1] - u[i*nx+j-1]) / (2.0 * dx)) * ((u[i*nx+j+1] - u[i*nx+j-1]) / (2.0 * dx)) -
            2.0 * ((u[(i+1)*nx+j] - u[(i-1)*nx+j]) / (2.0 * dy) *
              (v[i*nx+j+1] - v[i*nx+j-1]) / (2.0 * dx)) -
            ((v[(i+1)*nx+j] - v[(i-1)*nx+j]) / (2.0 * dy)) * ((v[(i+1)*nx+j] - v[(i-1)*nx+j]) / (2.0 * dy))));
    }
  }
}

void pressure_poisson(float *p, float *b){
  float pn[MAT_SIZE];

  for (int q = 0; q < nit; q++) {
    for (int i = 1; i < ny-1; i++) {
      for (int j = 1; j < nx-1; j++) {
        pn[i*nx+j] = p[i*nx+j];
      }
    }

    for (int i = 1; i < ny-1; i++) {
      for (int j = 1; j < nx-1; j++) {
        p[i*nx+j] = (((pn[i*nx+j+1] + pn[i*nx+j-1]) * pow(dy, 2) +
              (pn[(i+1)*nx+j] + pn[(i-1)*nx+j]) * pow(dx, 2)) /
            (2 * (pow(dx, 2) + pow(dy, 2))) -
            pow(dx, 2) * pow(dy, 2) / (2 * (pow(dx, 2) + pow(dy, 2))) *
            b[i*nx+j]);
      }
    }

    for (int i = 0; i < ny; i++) {
      p[i*nx+nx-1] = p[i*nx+nx-2];
      p[i*nx] = p[i*nx+1];
    }

    for (int i = 0; i < nx; i++) {
      p[i] = p[nx+i];
      p[nx*(ny-1)+i] = 0.0;
    }
  }
}

void cavity_flow(float *u, float *v, float *p) {
  float un[MAT_SIZE], vn[MAT_SIZE], b[MAT_SIZE];

  for (int i = 0; i < MAT_SIZE; i++) {
    un[i] = u[i];
    vn[i] = v[i];
    b[i] = 0.0;
  }

  for (int n = 0; n < nt; n++) {
    for (int i = 0; i < MAT_SIZE; i++) {
      un[i] = u[i];
      vn[i] = v[i];
    }

    build_up_b(u, v, b);
    pressure_poisson(p, b);

    for (int i = 1; i < ny-1; i++) {
      for (int j = 1; j < nx - 1; j++) {
        u[i*nx+j] = (un[i*nx+j] -
          un[i*nx+j] * dt / dx *
          (un[i*nx+j] - un[i*nx+j-1]) -
          vn[i*nx+j] * dt / dy *
          (un[i*nx+j] - un[(i-1)*nx+j]) -
          dt/ (2.0 * rho * dx) * (p[i*nx+j+1] - p[i*nx+j-1]) +
          nu * (dt / pow(dx, 2.0) *
            (un[i*nx+j+1] - 2.0 * un[i*nx+j] + un[i*nx+j-1]) +
            dt / pow(dy, 2.0) *
            (un[(i+1)*nx+j] - 2.0 * un[i*nx+j] + un[(i-1)*nx+j])));

        v[i*nx+j] = (vn[i*nx+j] -
            vn[i*nx+j] * dt / dx *
            (vn[i*nx+j] - vn[i*nx+j-1]) -
            vn[i*nx+j] * dt / dy *
            (vn[i*nx+j] - vn[(i-1)*nx+j]) -
            dt / (2 * rho * dy) * (p[(i+1)*nx+j] - p[(i-1)*nx+j]) +
            nu * (dt / pow(dx, 2) *
              (vn[i*nx+j+1] - 2 * vn[i*nx+j] + vn[i*nx+j-1]) +
              dt / pow(dy, 2) *
              (vn[(i+1)*nx+j] - 2.0 * vn[i*nx+j] + vn[(i-1)*nx+j])));
      }
    }

    for (int i = 0; i < nx; i++) {
      u[i] = 0.0;
      u[(ny-1)*nx+i] = 1.0;
      v[i] = 0.0;
      v[(ny-1)*nx+i] = 0.0;
    }

    for (int i = 0; i < ny; i++) {
      u[i*nx] = 0.0;
      u[i*nx+nx-1] = 0.0;
      v[i*nx] = 0.0;
      v[i*nx+nx-1] = 0.0;
    }

  }
}

int main() {
  float u[MAT_SIZE], v[MAT_SIZE], p[MAT_SIZE];

  for (int i = 0; i < MAT_SIZE; i++) {
    u[i] = 0.0;
    v[i] = 0.0;
    p[i] = 0.0;
  }

  cavity_flow(u, v, p);

  writeFile(u, v, p);

  return 0;
}
